base/experiments/ad9910_ram.py

```diff
@@ -0,0 +1,206 @@
+            case _:
+                raise NotImplementedError
+
+        self.start_addr = 0
+        self.end_addr = len(ram) - 1    # Inclusive
+
+        # Configure ther use of OSK. Note that using different OSK settings for
+        # different channels introduces step swiching latencies
+        if ram_type == RAMType.AMP or ram_type == RAMType.POLAR:
+            # Disable OSK. OSK amplitude as it has a higher priority than RAM.
+            self.osk_enable = 0
+        else:
+            # Enable OSK. There are no other utilizable amplitude data source.
+            self.osk_enable = 1
+
+        # Conversion to list to avoid type mismatch in the kernel
+        self.ram = ram.tolist()
+
+        # Note: Integer conversion may cause inaccuracy
+        self.step = int(ramp_interval * dds.sysclk / 4.0)
+
+        self.ramp_mode = ramp_mode
+
+
+class RAMProfileMap:
+    def __init__(self, core):
+        """Initialize a RAM profile to DDS mapping
+
+        Args:
+            core: The ARTIQ Core instance for time control.
+        """
+        self.ram_profile_map = []
+        self.cplds = []
+        self.core = core
+
+    def append(self, dds, ram_profile):
+        """Append a RAM profile into the builder. In addition, register the
+            CPLDs that have DDSes that playback the RAM profile.
+
+        Args:
+            dds: The DDS that will playback the RAM profile.
+            ram_profile: The RAM profile.
+        """
+        self.ram_profile_map.append((dds, ram_profile))
+
+        # Add the associlated CPLD to the list if not already there.
+        # ARTIQ-Python does not support iterration of a Python builtin set
+        # The workaround is to maintain a list, but convert it into set to
+        # avoid CPLD duplications.
+        cplds = set(self.cplds)
+        cplds.add(dds.cpld)
+        self.cplds = list(cplds)
+
+    @kernel
+    def program(self):
+        """Program the RAM profiles to the corresponding AD9910s.
+
+        """
+        # Switch to profile 0 for RAM. This is the default profile.
+        # Using other profiles for RAM is possible with appropriate parameters
+        # for future AD9910 function calls.
+        for cpld in self.cplds:
+            cpld.set_profile(0)
+
+        for dds, ram_profile in self.ram_profile_map:
+            # Datasheets strongly recommands setting ram_enable=0 before
+            # writing anything to the RAM profiles
+            dds.set_cfr1(ram_enable=0)
+            dds.cpld.io_update.pulse_mu(8)
+
+            # Write RAM profile that corresponds to the DDS
+            dds.set_profile_ram(
+                start=ram_profile.start_addr,
+                end=ram_profile.end_addr,
+                step=ram_profile.step,
+                mode=ram_profile.ramp_mode)
+            dds.cpld.io_update.pulse_mu(8)
+
+            # Program the RAM, break_realtime to avoid RTIOunderflow
+            dds.write_ram(ram_profile.ram)
```

Comment:
How long does this roughly take? Maybe document this in the wiki part of the jax.

Reply:
Tested by commenting out everything except `write_ram()` and the for loop that encapsulates it, with 1 RAM profile, 1024 entries.
Program:
```python
    @kernel
    def run(self):
        self.core.break_realtime()
        self.profile_map.load_ram()
```
```
(jlab) C:\Users\scientist\code\artiq>artiq_coreanalyzer -p
Log channel: 25
DDS one-hot: True
OutputMessage(channel=12, timestamp=1456682515160, rtio_counter=1456682396088, address=1, data=67110664)
OutputMessage(channel=12, timestamp=1456682515168, rtio_counter=1456682397152, address=0, data=369098752)
OutputMessage(channel=12, timestamp=1456682515320, rtio_counter=1456682398072, address=1, data=67116808)
OutputMessage(channel=12, timestamp=1456682515328, rtio_counter=1456682399952, address=0, data=4294705152)
OutputMessage(channel=12, timestamp=1456682515864, rtio_counter=1456682401080, address=0, data=4290510848)
...
OutputMessage(channel=12, timestamp=1456683063120, rtio_counter=1456683033728, address=0, data=4194304)
OutputMessage(channel=12, timestamp=1456683063656, rtio_counter=1456683034984, address=1, data=67116810)
OutputMessage(channel=12, timestamp=1456683063664, rtio_counter=1456683036360, address=0, data=0)
StoppedMessage(rtio_counter=1459112239920)
```
The entire call takes around 640 us to put into the FIFO, the `rtio_counter` and `timestamp` gap goes from around 120 us to around 27 us. It actually doesn't underflow on its own, but adding one more RAM profile could.

Reply:
Can you add the approximate timings into the docstring of this function? I assume you tested on Kasli 2.0 (I don't know if it would be the same on the Kasli-SoC).

Reply:
Added to docstring.